src/runtime/wasm.ts

```diff
@@ -175,113 +175,223 @@ class StringDecoder {
+   * any fields needed on wasmParticle, such as memory, tables, etc.
+   */
+  configureEnvironment(module: WebAssembly.Module, wasmParticle: WasmParticle, env: {});
+
+  /**
+   * Initializes the instantiated WebAssembly, runs any startup lifecycle, and if this
+   * runtime manages its own memory initialization, initializes the heap pointers on
+   * WasmParticle.
+   */
+  initializeInstance(wasmParticle: WasmParticle, instance: WebAssembly.Instance);
+}
+
+class EmscriptenWasmDriver implements WasmDriver {
+  shouldRun(module: WebAssembly.Module):boolean {
+      const customSections = WebAssembly.Module.customSections(module, 'emscripten_metadata');
+      if (customSections.length === 1) {
+          return true;
+      }
+      return false;
+  }
+
+  readEmscriptenMetadata(module: WebAssembly.Module) {
+    // Wasm modules built by emscripten require some external memory configuration by the caller,
+    // which is usually built into the glue code generated alongside the module. We're not using
+    // the glue code, but if we set the EMIT_EMSCRIPTEN_METADATA flag when building, emscripten
+    // will provide a custom section in the module itself with the required values.
+    const EMSCRIPTEN_METADATA_MAJOR = 0;
+    const EMSCRIPTEN_METADATA_MINOR = 1;
+    const EMSCRIPTEN_ABI_MAJOR = 0;
+    const EMSCRIPTEN_ABI_MINOR = 3;
+
+    const customSections = WebAssembly.Module.customSections(module, 'emscripten_metadata');
+    assert(customSections.length === 1, 'wasm particles must be built with EMIT_EMSCRIPTEN_METADATA');
+    if (customSections.length === 0) {
+      return null;
+    }
 
-// Wasm modules built by emscripten require some external memory configuration by the caller,
-// which is usually built into the glue code generated alongside the module. We're not using
-// the glue code, but if we set the EMIT_EMSCRIPTEN_METADATA flag when building, emscripten
-// will provide a custom section in the module itself with the required values.
-const EMSCRIPTEN_METADATA_MAJOR = 0;
-const EMSCRIPTEN_METADATA_MINOR = 1;
-const EMSCRIPTEN_ABI_MAJOR = 0;
-const EMSCRIPTEN_ABI_MINOR = 3;
-
-// TODO: reconcile with Kotlin-based particles
-function readEmscriptenMetadata(module: WebAssembly.Module) {
-  const customSections = WebAssembly.Module.customSections(module, 'emscripten_metadata');
-  assert(customSections.length === 1, 'wasm particles must be built with EMIT_EMSCRIPTEN_METADATA');
-
-  const buffer = new Uint8Array(customSections[0]);
-  const metadata: number[] = [];
-  let offset = 0;
-  while (offset < buffer.byteLength) {
-    let result = 0;
-    let shift = 0;
-    while (1) {
-      const byte = buffer[offset++];
-      result |= (byte & 0x7f) << shift;
-      if (!(byte & 0x80)) {
-        break;
+    const buffer = new Uint8Array(customSections[0]);
+    const metadata: number[] = [];
+    let offset = 0;
+    while (offset < buffer.byteLength) {
```

Comment:
What is this logic doing, can we document it a bit?

Reply:
This is just a refactor of the C++ code. 

Emscripten and Kotlin differ in memory management. Emscripten is more low level, you create the heap/WebAssembly.Memory yourself, but how should you size it? Emscripten provides metadata compiled into the WASM file for this purpose. There's also some expected heap pointers poked into the memory. 

Kotlin on the other hand, handles everything for you, so all you need to do is fetch what the heap created was.